### PR TITLE
Add user info in top queries

### PR DESCRIPTION
**Pull request by ansjcy**
### Description
Add remote ip address for top n queries.

### Related Issues
This is a follow up of https://github.com/opensearch-project/OpenSearch/pull/11904 to finish https://github.com/opensearch-project/OpenSearch/issues/11186

### tests
- Start an OpenSearch Cluster.
- Install `query-insights` plugin and enable top n queries
```
curl -X PUT 'localhost:9200/_cluster/settings' -H 'Content-Type: application/json' -d'
{
    "persistent" : {
        "search.insights.top_queries.latency.enabled" : "true",
        "search.insights.top_queries.latency.window_size" : "1m",
        "search.insights.top_queries.latency.top_n_size" : 5
    }
}'
```
- Index a document for searching
```
POST /my-index-0/_doc/?pretty
{
  "@timestamp": "2099-11-15T13:12:00",
  "message": "this is document 1",
  "user": {
    "id": "cyji"
  }
}
```
- Do `search`  on different browsers
```
GET _search
{
  "query": {
    "match_all": {}
  }
}
```
- Check top queries
```
GET _insights/top_queries

{
  "top_queries": [
    {
      "timestamp": 1709161455119,
      "search_type": "query_then_fetch",
      "node_id": "RTr6rGKbSW-9_Lmli2HVYA",
      "total_shards": 10,
      "indices": [],
      "phase_latency_map": {
        "expand": 0,
        "query": 7,
        "fetch": 5
      },
      "source": "{\"query\":{\"match_all\":{\"boost\":1.0}}}",
      "remote_address": "172.18.0.2:37416",
      "latency": 14
    },
    {
      "timestamp": 1709161458352,
      "search_type": "query_then_fetch",
      "node_id": "yNwke7ByTZOsdSsIaGZdbw",
      "total_shards": 10,
      "indices": [],
      "phase_latency_map": {
        "expand": 0,
        "query": 7,
        "fetch": 3
      },
      "source": "{\"query\":{\"match_all\":{\"boost\":1.0}}}",
      "remote_address": "172.18.0.2:37412",
      "latency": 12
    },
    {
      "timestamp": 1709161464149,
      "search_type": "query_then_fetch",
      "node_id": "RTr6rGKbSW-9_Lmli2HVYA",
      "total_shards": 10,
      "indices": [],
      "phase_latency_map": {
        "expand": 0,
        "query": 6,
        "fetch": 2
      },
      "source": "{\"query\":{\"match_all\":{\"boost\":1.0}}}",
      "remote_address": "172.18.0.2:37416",
      "latency": 10
    }
  ]
}
```

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ansjcy** [comment]
We shouldn't add dependencies on common-utils before we move queries insights to its own repo. Otherwise it will introduce circular dependency between OpenSearch core repo and common-utils repo.

File: plugins/query-insights/src/main/java/org/opensearch/plugin/insights/utils/ThreadContextParser.java
Comment (sandeshkr419) [inline plugins/query-insights/src/main/java/org/opensearch/plugin/insights/utils/ThreadContextParser.java:23]: Should we make this class as static?

**Comment by ansjcy** [comment]
Discussed with .peternied, reading from thread context is actually not recommended. We should explore extending `IdentityPlugin` https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/IdentityPlugin.java#L19 for this purpose.

**Review by peternied — CHANGES_REQUESTED** [review]
@ansjcy Thanks for creating this PR - it sounds like a really useful feature for sysadmins.

This is a great use case for Identity - where information about who is operating a request should come from.  You can use the other plugin identity-shiro [1] to prove out your scenarios.

Furthermore there is a problem of 'userless' queries.  There are many system tasks and plugin operations via job scheduler that will not have user identity tied to them.  While not directly related if pulled from identity systems, the lack of insight here might warrant follow up work for your use case.  

If this PR is slimed down the the source - ip address that seems like a iterative step while building up support for exposing user identity.

- [1] https://github.com/opensearch-project/OpenSearch/tree/main/plugins/identity-shiro

**Review by peternied — COMMENTED** [review]
Is there an integration test for this scenario?  If not lets add one, or edit the existing test case to verify remote addresses are sensible.

File: plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/listener/QueryInsightsListener.java
Comment (peternied) [reply plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/listener/QueryInsightsListener.java:141]: Thanks for the detail. This might be an area to look into if it becomes cumbersome... such as the changes that were added for the unit test 😉

**Review by peternied — COMMENTED** [review]
I am about to go on vacation so I won't be able to follow up for a while.

Fellow maintainers please feel free to dismiss my review when integration tests are added/updated to unblock merging of this PR. Thanks!

File: server/src/main/java/org/opensearch/rest/action/search/RestSearchAction.java
Comment (msfroh) [inline server/src/main/java/org/opensearch/rest/action/search/RestSearchAction.java:229]: I don't think it's safe to hijack `remoteAddress()` like this. 

It's intended to be set by `InboundHandler` on transport requests. Setting it on a REST request may have unintended consequences.

Also, if the request comes through a load balancer or a proxy, this value will almost certainly be the load balancer or proxy's address.

**Comment by ansjcy** [comment]
Closing the PR as per peternied and froh's comments. We need to again carefully review what should be the best way to move forward to add user related information to top queries - Let's check the Identity service to see what are the gaps to get user info from there.

**Pull request closed without merging**
